### PR TITLE
[MPS] Implement true adaptive max pooling

### LINKED ISSUE #190064: [MPS] adaptive_max_pool2d returns wrong values, indices, and even wrong output shape for non-divisible geometry
### 🐛 Describe the bug

MPS implements `adaptive_max_pool2d` as a single ordinary max-pool with one
derived kernel/stride pair, without requiring divisibility. True adaptive
pooling uses windows of varying size that can overlap, so for non-divisible
input/output geometry the values and indices are wrong — and for upsampling
geometry the output even has the wrong shape.

```python
import torch

x = torch.arange(25.0).reshape(1, 1, 5, 5)
ref, ref_idx = torch.nn.functional.adaptive_max_pool2d(x, (3, 3), return_indices=True)
out, idx = torch.nn.functional.adaptive_max_pool2d(x.to("mps"), (3, 3), return_indices=True)
print("cpu values :", ref.flatten().tolist())
print("mps values :", out.cpu().flatten().tolist())
print("cpu indices:", ref_idx.flatten().tolist())
print("mps indices:", idx.cpu().flatten().tolist())
up = torch.nn.functional.adaptive_max_pool2d(torch.randn(1, 1, 2, 2, device="mps"), (4, 4))
print("2x2 -> 4x4:", tuple(up.shape))
```

Output on a source build of current main (2.14.0a0+git2662808); also
reproduces on 2.13.0 stable:

```text
cpu values : [6.0, 8.0, 9.0, 16.0, 18.0, 19.0, 21.0, 23.0, 24.0]
mps values : [12.0, 13.0, 14.0, 17.0, 18.0, 19.0, 22.0, 23.0, 24.0]
cpu indices: [6, 8, 9, 16, 18, 19, 21, 23, 24]
mps indices: [12, 13, 14, 17, 18, 19, 22, 23, 24]
2x2 -> 4x4: (1, 1, 1, 1)
```

The last line should be `(1, 1, 4, 4)`; a deprecation warning about resizing
the out tensor is also emitted on stderr. Backward inherits the same
incorrect geometry through the indices.

The implementation derives one fixed kernel/stride and delegates to
`max_pool2d_with_indices`:
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/operations/AdaptivePooling.mm#L182-L214
Note the sibling `adaptive_avg_pool2d` at least raises for non-divisible
sizes (#96056); `adaptive_max_pool2d` silently returns wrong results.

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main) and 2.13.0; macOS 26.5.2 (arm64), Apple M5 Max, Python 3.12.13.

cc @ezyang @gchanan @kadeng @msaroufim @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by bogatyy**
`adaptive_max_pool2d` was implemented as a single ordinary max-pool with one derived kernel/stride pair, without requiring divisibility. True adaptive pooling uses windows `[floor(o*I/O), ceil((o+1)*I/O))` that vary in size and overlap for non-divisible geometry, so values and indices were silently wrong (5×5 → 3×3 returned `[12,13,14,...]` where CPU returns `[6,8,9,...]`), upsampling geometry returned the wrong output shape entirely (2×2 → 4×4 produced 1×1), and backward inherited the bad indices.

This adds dedicated Metal kernels: forward computes one output element per thread with the adaptive window bounds, propagating NaN with first-NaN-wins index semantics like CPU; backward atomically routes grad_output through the saved indices (windows overlap, so multiple outputs can hit one input element). Input offsets use 64-bit math since the input can exceed 2^31 elements even when the output is small. The host impls move to `Pooling.mm` next to the kernels' metallib; non-contiguous tensors go through contiguous temporaries.

`test_adaptive_max_pool2d_simple` gains non-divisible, channels-last, and upsampling geometries (values, indices, and gradients against CPU), plus a new NaN-propagation test. All 20 adaptive and 117 pool tests pass.

**Alternatives considered**

- *Raise for non-divisible sizes, like `adaptive_avg_pool2d` does (~3 lines):* stops the silent corruption but turns the op's core use case (arbitrary input sizes) into an error; the avg-pool version of that error is itself a long-standing complaint (#96056). Defensible only as a stopgap.
- *Compose from existing ops:* adaptive window starts follow `floor(o*I/O)`, an irregular pattern with two window sizes that can overlap, so no fixed (kernel, stride, padding, dilation) reproduces them. A host-side loop over output cells needs O(H_out*W_out) dispatches and still has to build the flattened indices tensor for autograd.
- *Custom forward, reusing `max_pool2d_with_indices_backward`:* the MPS graph path of that backward recomputes the argmax from the supplied kernel geometry instead of using the saved indices, so fabricated geometry would reintroduce the same bug class.

Hence dedicated forward/backward kernels.

Fixes #190064

*This PR was authored with assistance from Claude.*

### CODE REVIEW

**Comment by bogatyy** [comment]
Closing in favor of #189659, which predates this PR and fixes the same root cause with a cleaner integration: an adaptive flag on the existing stride-aware max_pool kernel instead of dedicated kernels, with native channels_last support and the existing backward reused. The one gap, NaN propagation in the shared kernel, is flagged over there.
